src/qibo/backends/pytorch.py

```diff
@@ -0,0 +1,223 @@
+        else:
+            x = self.np.tensor(x, dtype=dtype)
+
+        if copy:
+            return x.clone()
+
+        return x
+
+    def issparse(self, x):
+        if isinstance(x, self.np.Tensor):
+            return x.is_sparse
+
+        return super().issparse(x)
+
+    def to_numpy(self, x):
+        if isinstance(x, list):
+            return np.asarray([self.to_numpy(i) for i in x])
+
+        if isinstance(x, self.np.Tensor):
+            return x.numpy(force=True)
+
+        return x
+
+    def _append_zeros(self, state, qubits, results):
+        """Helper method for collapse."""
+        for q, r in zip(qubits, results):
+            state = self.np.unsqueeze(state, dim=q)
+            if r:
+                state = self.np.cat([self.np.zeros_like(state), state], dim=q)
+            else:
+                state = self.np.cat([state, self.np.zeros_like(state)], dim=q)
+        return state
+
+    def _order_probabilities(self, probs, qubits, nqubits):
+        """Arrange probabilities according to the given ``qubits`` ordering."""
+        if probs.dim() == 0:  # pragma: no cover
+            return probs
+        unmeasured, reduced = [], {}
+        for i in range(nqubits):
+            if i in qubits:
+                reduced[i] = i - len(unmeasured)
+            else:
+                unmeasured.append(i)
+        return self.np.transpose(probs, [reduced.get(i) for i in qubits])
+
+    def calculate_probabilities(self, state, qubits, nqubits):
+        rtype = self.np.real(state).dtype
+        unmeasured_qubits = tuple(i for i in range(nqubits) if i not in qubits)
+        state = self.np.reshape(self.np.abs(state) ** 2, nqubits * (2,))
+        if len(unmeasured_qubits) == 0:
+            probs = self.cast(state, dtype=rtype)
+        else:
+            probs = self.np.sum(self.cast(state, dtype=rtype), axis=unmeasured_qubits)
+        return self._order_probabilities(probs, qubits, nqubits).ravel()
+
+    def set_seed(self, seed):
+        self.np.manual_seed(seed)
+        np.random.seed(seed)
+
+    def sample_shots(self, probabilities, nshots):
+        return self.np.multinomial(
+            self.cast(probabilities, dtype="float"), nshots, replacement=True
+        )
+
+    def samples_to_binary(self, samples, nqubits):
+        samples = self.cast(samples, dtype="int32")
+        qrange = self.np.arange(nqubits - 1, -1, -1, dtype=self.np.int32)
+        samples = samples.int()
+        samples = samples[:, None] >> qrange
+        return samples % 2
+
+    def samples_to_decimal(self, samples, nqubits):
+        samples = self.cast(samples, dtype="int32")
+        qrange = self.np.arange(nqubits - 1, -1, -1, dtype=torch.int32)
+        qrange = (2**qrange).unsqueeze(1)
+        return self.np.matmul(samples, qrange).squeeze(1)
+
+    def calculate_overlap_density_matrix(self, state1, state2):
+        return self.np.trace(
+            self.np.matmul(self.np.conj(self.cast(state1)).T, self.cast(state2))
+        )
```

Comment:
I tried to remove this method and tests started failing. My guess is that the issue is due to `self.np.transpose` in the `torch` backend being overwritten by `torch.permute` because of other methods.

Reply:
I solved this using the .T method that works for both backends